nixos/postgresql: support 0750 for data directory

danbst opened the pull request:
This is rework of part of https://github.com/NixOS/nixpkgs/pull/46670.
My usecase was to be able to inspect PG datadir as wheel user.

PG11 now allows starting server with 0750 mask for data dir.
`groupAccess = true` now does this automatically. The only thing you have to do
is to set group ownership.

For PG10 and below, I've described a hack how this can be done. Before this PR
hack was impossible. The hack isn't ideal, because there is short
period of time when dir mode is 0700, so I didn't want to make it official.

Test/example is present too.

danbst committed:
nixos/postgresql: support 0750 for data directory

This is rework of part of https://github.com/NixOS/nixpkgs/pull/46670.
My usecase was to be able to inspect PG datadir as wheel user.

PG11 now allows starting server with 0750 mask for data dir.
`groupAccess = true` now does this automatically. The only thing you have to do
is to set group ownership.

For PG10 and below, I've described a hack how this can be done. Before this PR
hack was impossible. The hack isn't ideal, because there is short
period of time when dir mode is 0700, so I didn't want to make it official.

Test/example is present too.

danbst commented:
Does it look fine to mention hack in docs in option description?

danbst committed:
postgresql: allow changing initidb arguments via module system

Closes https://github.com/NixOS/nixpkgs/issues/18829

+ some cleanups

danbst commented:
Added a commit to fix https://github.com/NixOS/nixpkgs/issues/18829

danbst commented:
requesting also review from @pacien

pacien commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -80,6 +82,23 @@ in
+      groupAccess = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
```
Nitpick: unnecessary example for boolean option.

pacien commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -80,6 +82,23 @@ in
+        description = ''
+          Allow read access for group (0750 mask for data directory).
+          Supported only for PostgreSQL 11+. PostgreSQL 10 and lower doesn't
+          support starting server with 0750 mask, but a workaround like
```
IMHO, we shouldn't encourage such hacks in the official documentation.
This workaround for PostgreSQL <= 10 should probably go on the wiki instead.

pacien commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -249,8 +285,12 @@ in
+                  else "/var/db/postgresql");
+
+    services.postgresql.initdbFlags =
+      mkDefault (lib.optional cfg.groupAccess "--allow-group-access");
```
Setting both `groupAccess = true` and `initdbFlags = [ "--some-other-flag" ]` causes `"--allow-group"` not to be added due to `mkDefault`.

It's simpler and more explicit to combine arguments when calling `initdb` in `script` than making the latter option depend on the former.

danbst replied:
Indeed, thanks! I really shoud use `mkBefore` instead:
```
# nix repl ./lib
nix-repl> (evalModules { modules = [ 
                   { options.bla = mkOption { type = with types; listOf int; }; 
                     config.bla = mkBefore [1]; 
                   } 
                   { bla = [2];  } 
               ]; }).config.bla
[ 1 2 ]
```
Will change test.

pacien commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -287,17 +327,17 @@ in
             # Initialise the database.
             if ! test -e ${cfg.dataDir}/PG_VERSION; then
-              initdb -U ${cfg.superUser}
+              initdb -U ${cfg.superUser} ${lib.concatStringsSep " " cfg.initdbFlags}
```
Combining flags here is more explicit.
Nitpick: `lib.` is also not necessary.

```suggestion
              initdb -U ${cfg.superUser} \
                ${optionalString cfg.groupAccess "--allow-group-access"} \
                ${concatStringsSep " " cfg.initdbFlags}
```

danbst replied:
I see your point. I'll stick to module-system style though, it's not that bad.

I'll remove `lib.`

pacien commented on nixos/tests/postgresql.nix:
```diff
@@ -69,5 +69,71 @@ let
+      $machine->succeed("echo select 1 | sudo -u postgres psql"); # works after restart
+      $machine->succeed("sudo -u backup ls ${dataDir}");
+
+      # This tests a hack for PG <11: restore permissions to 0700 just before PG starts
```
I'm not sure whether we have to test this workaround.

pacien commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -306,8 +346,9 @@ in
               ln -sfn "${pkgs.writeText "recovery.conf" cfg.recoveryConfig}" \
                 "${cfg.dataDir}/recovery.conf"
             ''}
+            chmod ${dirMode} "${cfg.dataDir}"
```
Is it necessary to re-apply the mode at each startup?
If so, this should perhaps be applied recursively.
This might be a costly operation on big databases though.

danbst replied:
PostgreSQL is sensitive only for group mode of dataDir itself. Files inside can be even world readable, but top-level dataDir may be only 0700 or 0750 (for pg11). Otherwise it just doesn't start.

danbst replied:
@pacien ok, I made it optional. Now `groupAccess` is tristate with `null` being "don't change mask" (previous behavior).

pacien replied:
How does PostgreSQL apply the right permissions to new files within the data directory?
Is it copying the mode of dataDir itself?

danbst committed:
addressed review comments and some fixes

danbst committed:
whoops

danbst committed:
change groupAccess to tristate, to not force `chmod` on dataDir.

Making mask either 0700 or 0750 is too restrictive..

spinus commented on nixos/tests/postgresql.nix:
```diff
@@ -69,5 +69,47 @@ let
+      machine = { config, ...}:
+        {
+          services.postgresql.enable = true;
+          services.postgresql.package = pkgs.postgresql_11;
```
why not to make test for all the versions with generator?

danbst replied:
This feature isn't supported for PG < 11. So not sure there is need to test it for 11, 12, 13 only (last two not yet released).

spinus commented on nixos/tests/postgresql.nix:
```diff
@@ -69,5 +69,47 @@ let
   (mapAttrs' (name: package: { inherit name; value=make-postgresql-test name package false;}) postgresql-versions) // {
     postgresql_11-backup-all = make-postgresql-test "postgresql_11-backup-all" postgresql-versions.postgresql_11 true;
+
+    postgresql_dirmode_change =
```
+1 for updating tests

danbst replied:
I had to remove it, as I don't want to support old pgs, and new ones work fine.

danbst committed:
WIP

danbst committed:
Merge branch 'master' into postgresql_group

danbst committed:
let's not support group mode for versions pre-11.

The only fix is to change mode to 0700 before start, because otherwise postgresql
doesn't start, and error is non-obvious.

danbst commented:
I've rethought my approach. Let's not support pre11 versions THAT much. It is enough to just ensure state dir has 0700 mode for old PG.

So now dir mode is enforced only when PG is older than 11. This fixes a nasty issue when you add tmpfiles rule to make PG group-accessible and suddenly on next reboot PG doesn't start (yes, I had this...).

Now all 3 examples below start postgres, and last are possible ways to set group mode.
```
{
  documentation.nixos.enable = false;

  containers.test1.config = {config, pkgs, ...}: 
        {
          services.postgresql.enable = true;
          services.postgresql.port = 10000;
          services.postgresql.package = pkgs.postgresql_10;
          systemd.tmpfiles.rules = [ ''d "${config.services.postgresql.dataDir}" 0750 - - -''  ];
        };
  containers.test2.config = {config, ...}: 
        {
          services.postgresql.enable = true;
          services.postgresql.port = 10000;
          systemd.tmpfiles.rules = [ ''d "${config.services.postgresql.dataDir}" 0750 - - -''  ];
        };
  containers.test3.config = {config, ...}: 
        {
          services.postgresql.enable = true;
          services.postgresql.port = 10001;
          services.postgresql.initdbArgs = [ "--allow-group-access" ];
        };
}
```

danbst commented:
urgh. Github UI discarded my perfect squash-commit message...

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -88,6 +90,16 @@ in
+        default = [];
+        example = [ "--data-checksums" "--allow-group-access" ];
+        description = ''
+          Additional arguments passed to <literal>initdb<literal> during data dir
```
For the record, this breaks the manual build: 6c6310787212acaba1b8f3a9b3bfd319015f00e2

danbst replied:
:-( thanks for the fix
